sui-execution/latest/sui-move-natives/src/crypto/sui_state_proof.rs

```diff
@@ -308,17 +339,34 @@ pub fn sui_state_proof_verify_transaction(
         return Ok(NativeResult::err(cost, INVALID_TX));
     }
 
-    // Ensure the tx is part of the checkpoint
-    let is_valid_checkpoint_tx = checkpoint_contents
-        .iter()
-        .any(|&digest| digest == transaction.effects.execution_digests());
-    if !is_valid_checkpoint_tx {
+    let digests = transaction.effects.execution_digests();
+
+    // Check that transaction digest matches the execution digest
+    if transaction.transaction.digest() != &digests.transaction {
+        return Ok(NativeResult::err(cost, INVALID_TX));
+    }
+
+    // Ensure the digests are in the checkpoint contents
+    if !checkpoint_contents
+        .enumerate_transactions(&summary)
+        .any(|x| x.1 == &digests) {
+            return Ok(NativeResult::err(cost, INVALID_TX));
+    }   
+
+    let events_digest = transaction.events.as_ref().map(|events| events.digest());
+    // Ensure that the execution digest matches the events digest of the passed transaction
+    if !(events_digest.as_ref() == transaction.effects.events_digest()) {
         return Ok(NativeResult::err(cost, INVALID_TX));
     };
 
-    let tx_events = &transaction.events.as_ref().unwrap().data;
+    let tx_events = match transaction.events.as_ref() {
+        Some(events) => &events.data,
+        None => {
+            return Ok(NativeResult::err(cost, INVALID_TX));
+        }
+    };
```

Comment:
This code is duplicated in both `sui_state_proof_verify_transaction()` and `sui_state_proof_verify_link_cap()` please DRY it by moving it to a function.

Reply:
Introduced a helper fun